flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -1479,6 +1485,72 @@ public void testLateRegisteredPartitionsWillBeReleased() {
+        final JobGraph jobGraph = singleJobVertexJobGraph(1);
+        enableCheckpointing(jobGraph);
+        final CountDownLatch checkpointServicesShutdownBlocked = new CountDownLatch(1);
+        final CountDownLatch cleanerClosed = new CountDownLatch(1);
+        final ScheduledExecutorService executorService =
+                Executors.newSingleThreadScheduledExecutor();
+        final CompletedCheckpointStore completedCheckpointStore =
+                new StandaloneCompletedCheckpointStore(1) {
+
+                    @Override
+                    public void shutdown(JobStatus jobStatus, CheckpointsCleaner checkpointsCleaner)
+                            throws Exception {
+                        checkpointServicesShutdownBlocked.await();
+                        super.shutdown(jobStatus, checkpointsCleaner);
+                    }
+                };
+        final CheckpointIDCounter checkpointIDCounter =
+                new StandaloneCheckpointIDCounter() {
+
+                    @Override
+                    public void shutdown(JobStatus jobStatus) throws Exception {
+                        checkpointServicesShutdownBlocked.await();
+                        super.shutdown(jobStatus);
+                    }
+                };
+        final DefaultScheduler scheduler =
+                SchedulerTestingUtils.newSchedulerBuilder(
+                                jobGraph,
+                                ComponentMainThreadExecutorServiceAdapter.forSingleThreadExecutor(
+                                        executorService))
+                        .setCheckpointRecoveryFactory(
+                                new TestingCheckpointRecoveryFactory(
+                                        completedCheckpointStore, checkpointIDCounter))
+                        .setCheckpointCleaner(
+                                new CheckpointsCleaner() {
+
+                                    @Override
+                                    public synchronized CompletableFuture<Void> closeAsync() {
+                                        cleanerClosed.countDown();
+                                        return super.closeAsync();
+                                    }
+                                })
+                        .build();
+
+        final CompletableFuture<Void> schedulerClosed = new CompletableFuture<>();
+        final CountDownLatch schedulerClosing = new CountDownLatch(1);
+
+        executorService.submit(
+                () -> {
+                    scheduler.closeAsync().thenRun(() -> schedulerClosed.complete(null));
+                    schedulerClosing.countDown();
+                });
+
+        // Wait for scheduler to start closing.
+        schedulerClosing.await();
+        assertFalse(
+                "CheckpointCleaner should not close before checkpoint services.",
+                cleanerClosed.await(10, TimeUnit.MILLISECONDS));
+        checkpointServicesShutdownBlocked.countDown();
```

Comment:
before counting down, you could check whether `cleanerClosed` has been counted down or not.

Reply:
That's already checked by the call above isn't it? By asserting that await has "timed out".

Reply:
True. Forget what I've said :-)